enterprise/coderd/identityprovider/tokens.go

```diff
@@ -0,0 +1,246 @@
+		return oauth2.Token{}, sql.ErrNoRows
+	}
+
+	// Generate a refresh token.
+	// The refresh token is not currently used or exposed though as API keys can
+	// already be refreshed by just using them.
+	// TODO: However, should we implement the refresh grant anyway?
+	// 40 characters matches the length of GitHub's client secrets.
+	rawRefreshToken, err := cryptorand.String(40)
+	if err != nil {
+		return oauth2.Token{}, err
+	}
+
+	// Generate the API key we will swap for the code.
+	// TODO: We are ignoring scopes for now.
+	tokenName := fmt.Sprintf("%s_%s_oauth_session_token", dbCode.UserID, app.ID)
+	key, sessionToken, err := apikey.Generate(apikey.CreateParams{
+		UserID:    dbCode.UserID,
+		LoginType: database.LoginTypeOAuth2ProviderApp,
+		// TODO: This is just the lifetime for api keys, maybe have its own config
+		//       settings. #11693
+		DefaultLifetime: defaultLifetime,
+		// For now, we allow only one token per app and user at a time.
+		TokenName: tokenName,
+	})
+	if err != nil {
+		return oauth2.Token{}, err
+	}
+
+	// Grab the user roles so we can perform the exchange as the user.
+	//nolint:gocritic // In the token exchange, there is no user actor.
+	roles, err := db.GetAuthorizationUserRoles(dbauthz.AsSystemRestricted(ctx), dbCode.UserID)
+	if err != nil {
+		return oauth2.Token{}, err
+	}
+	userSubj := rbac.Subject{
+		ID:     dbCode.UserID.String(),
+		Roles:  rbac.RoleNames(roles.Roles),
+		Groups: roles.Groups,
+		Scope:  rbac.ScopeAll,
+	}
+
+	// Do the actual token exchange in the database.
+	err = db.InTx(func(tx database.Store) error {
+		ctx := dbauthz.As(ctx, userSubj)
+		err = tx.DeleteOAuth2ProviderAppCodeByID(ctx, dbCode.ID)
+		if err != nil {
+			return xerrors.Errorf("delete oauth2 app code: %w", err)
+		}
+
+		// Delete the previous key, if any.
+		prevKey, err := tx.GetAPIKeyByName(ctx, database.GetAPIKeyByNameParams{
+			UserID:    dbCode.UserID,
+			TokenName: tokenName,
+		})
+		if err == nil {
+			err = tx.DeleteAPIKeyByID(ctx, prevKey.ID)
+		}
+		if err != nil && !errors.Is(err, sql.ErrNoRows) {
+			return xerrors.Errorf("delete api key by name: %w", err)
+		}
+
+		newKey, err := tx.InsertAPIKey(ctx, key)
+		if err != nil {
+			return xerrors.Errorf("insert oauth2 access token: %w", err)
+		}
+
+		hashed := Hash(rawRefreshToken, app.ID)
+		_, err = tx.InsertOAuth2ProviderAppToken(ctx, database.InsertOAuth2ProviderAppTokenParams{
+			ID:          uuid.New(),
+			CreatedAt:   dbtime.Now(),
+			ExpiresAt:   key.ExpiresAt,
+			RefreshHash: hashed[:],
+			AppSecretID: secret.ID,
+			APIKeyID:    newKey.ID,
+		})
+		if err != nil {
+			return xerrors.Errorf("insert oauth2 refresh token: %w", err)
+		}
+		return nil
+	}, nil)
+	if err != nil {
+		return oauth2.Token{}, err
+	}
+
+	return oauth2.Token{
+		AccessToken: sessionToken,
+		TokenType:   "Bearer",
+		// TODO: Exclude until refresh grant is implemented.
+		// RefreshToken: rawRefreshToken,
+		// Expiry:       key.ExpiresAt,
```

Comment:
I see why you do not have Expiry, but that usually means it never expires.

I think we should add a expiry and disable auto refreshing of these api keys.

Reply:
Oh yeah I was trying to see if omitting it would make Backstage stop trying to refresh, but it tries to anyway.  I wanted to merge this and work on refresh next but maybe I have to add it now.